### ISSUE TITLE
setup:upgrade fails on Mageneto_Indexer step

### ISSUE THREAD

**Issue by kamilw-ed** [magento/magento2#3329]
Hi!

I'm trying to build an automated deployment to our server with TeamCity and phing for Magento 2. Currently we're using CE version but soon we'll upgrade it to EE. Anyway, I have completed my set up but TC build fails when it comes to setup:upgrade command. All steps for the build are not important right now as my issue can be easily replicated on local (more on that in a moment).

In our BitBucket repository we keep only composer.json, composer.lock and custom extensions and themes. Magento core is fetched by composer. So, I do git clone of that repository, then I do composer install. I have a env.php file already so I copy it and then do the `$ php bin/magento setup:upgrade` which ends up with the following error message:

```
PHP Fatal error:  Cannot instantiate interface Magento\Framework\Indexer\ConfigInterface in /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php on line 73
PHP Stack trace:
PHP   1. {main}() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/bin/magento:0
PHP   2. Symfony\Component\Console\Application->run() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/bin/magento:25
PHP   3. Magento\Framework\Console\Cli->doRun() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/symfony/console/Symfony/Component/Console/Application.php:126
PHP   4. Symfony\Component\Console\Application->doRun() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/Console/Cli.php:49
PHP   5. Symfony\Component\Console\Application->doRunCommand() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/symfony/console/Symfony/Component/Console/Application.php:195
PHP   6. Symfony\Component\Console\Command\Command->run() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/symfony/console/Symfony/Component/Console/Application.php:874
PHP   7. Magento\Setup\Console\Command\UpgradeCommand->execute() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/symfony/console/Symfony/Component/Console/Command/Command.php:257
PHP   8. Magento\Setup\Model\Installer->installSchema() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php:71
PHP   9. Magento\Setup\Model\Installer->handleDBSchemaData() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/setup/src/Magento/Setup/Model/Installer.php:732
PHP  10. Magento\Setup\Model\Installer->getSchemaDataHandler() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/setup/src/Magento/Setup/Model/Installer.php:818
PHP  11. Magento\Setup\Model\Installer->createSchemaDataHandler() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/setup/src/Magento/Setup/Model/Installer.php:1171
PHP  12. Magento\Framework\ObjectManager\ObjectManager->create() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/setup/src/Magento/Setup/Model/Installer.php:867
PHP  13. Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/ObjectManager/ObjectManager.php:57
PHP  14. Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:82
PHP  15. Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:53
PHP  16. Magento\Framework\ObjectManager\ObjectManager->get() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:236
PHP  17. Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() /media/crucial240/Projects/mage2demo/code/test2/magento2demo/src/vendor/magento/framework/ObjectManager/ObjectManager.php:71
```

How to reproduce it.
- install Magento 2
- clone repository in other place
- do composer install
- copy env.php
- do the `php bin/magento setup:upgrade`

I created a short asciinema video that show all the steps. 

[![Magento 2 setup:upgrade error](https://asciinema.org/a/da0rfnk0halxse4p3py833o5q.png)](https://asciinema.org/a/da0rfnk0halxse4p3py833o5q)

I don't want to be forced to do setup:upgrade then `rm -rf var/cache` and then again setup:upgrade. 
I tried to inspect the code but so far I found nothing.  Please help me with this issue.

**Comment by mazhalai** [comment]
@kamilw-ed I don't understand why you need to do both composer create project as well as git clone. can you please provide more details regarding that?

Also, installing in another place `B` which probably has different modules than in location `A`, and then  copying `B/etc/env.php` to `A/etc/env.php` will not work.You need to copy etc/config.php too and also have the exact same modules in both places. If you used composer, it should be under `vendor/magento` and if you git clone it wil be under `/app/code`. 

This sentence is contradictory: 

```
Magento core is fetched by composer. So, I do git clone of that repository,
```

You can either do composer create-project or git clone, not both, so please clarify. 

Git clone is only recommended for [contributing developers, it is not meant to be run in production](http://devdocs.magento.com/guides/v2.0/install-gde/continue.html). Also, git clone gets the latest from develop branch which is WIP, the stable branch is 2.0.2 so you should be switching to that branch.

**Comment by kamilw-ed** [comment]
Thanks for reply @mazhalai, I guess I need to provide more details into this issue. 
I have created Magento 2 project using `composer create-project`, I followed all steps from Magento website. It downloaded the Magento stuff. I commited to our company's git repository only 2 files: composer.json and composer.lock. We decided to not keep Magento core files in the repo itself as they can be fetched by composer. So in place A on my disk I created the project, then installed it and got it working. 
To simulate our build and deployment strategy I created a folder at place B. I did a code checkout from the same repository (so it's using exact same modules and configuration). I did composer install and copy of `env.php` as part of the simulation. This is what our TeamCity would do. So, having both the code and configuration in place I'm ready to run `setup:upgrade` to enforce module installation and updates. Next step would be to do `setup:di:compile-multi-teant`, then `setup:static-content:deploy` and finally `cache:flush`. 
But I'm stuck on the `setup:upgrade` step because I'm getting that error message mentioned in my first post. 
I hope it's more clear now about what I'm trying to do here.

EDIT
I solved the issue. It appeared that I forgot to commit `app/etc/config.php` and that's why it was failing. Thanks for support. Ticket can be closed.
